grpc/public/public_server.go

```diff
@@ -0,0 +1,211 @@
+)
+
+// Compile-time check
+var _ authnpb.PublicAuthNServer = publicServer{}
+
+type publicServer struct {
+	app *app.App
+}
+
+func RunPublicGRPC(ctx context.Context, app *app.App, l net.Listener) error {
+	opts := meta.PublicServerOptions(app)
+	srv := grpc.NewServer(opts...)
+
+	RegisterPublicGRPCMethods(srv, app)
+
+	go func() {
+		<-ctx.Done()
+		srv.GracefulStop()
+	}()
+
+	if err := srv.Serve(l); err != nil {
+		log.Printf("serve error: %s", err)
+		return err
+	}
+	return nil
+}
+
+func RegisterPublicGRPCMethods(srv *grpc.Server, app *app.App) {
+	authnpb.RegisterPublicAuthNServer(srv, publicServer{
+		app: app,
+	})
+
+	if app.Config.EnableSignup {
+		authnpb.RegisterSignupServiceServer(srv, signupServiceServer{
+			app: app,
+		})
+	}
+
+	if app.Config.AppPasswordResetURL != nil {
+		authnpb.RegisterPasswordResetServiceServer(srv, passwordResetServer{
+			app: app,
+		})
+	}
+
+	if app.Config.AppPasswordlessTokenURL != nil {
+		authnpb.RegisterPasswordlessServiceServer(srv, passwordlessServer{
+			app: app,
+		})
+	}
+}
+
+func (s publicServer) Login(ctx context.Context, req *authnpb.LoginRequest) (*authnpb.LoginResponseEnvelope, error) {
+	account, err := services.CredentialsVerifier(
+		s.app.AccountStore,
+		s.app.Config,
+		req.GetUsername(),
+		req.GetPassword(),
+	)
+	if err != nil {
+		if fe, ok := err.(services.FieldErrors); ok {
+			br := &errdetails.BadRequest{}
+			for _, fe := range fe {
+				br.FieldViolations = append(br.FieldViolations, &errdetails.BadRequest_FieldViolation{
+					Field:       fe.Field,
+					Description: fe.Message,
+				})
+			}
+			statusError := status.New(codes.FailedPrecondition, fe.Error())
+			statusEr, e := statusError.WithDetails(br)
+			if e != nil {
+				panic(fmt.Sprintf("Unexpected error attaching metadata: %v", e))
+			}
+			return nil, statusEr.Err()
+		}
+		panic(err)
+	}
+
+	sessionToken, identityToken, err := services.SessionCreator(
+		s.app.AccountStore, s.app.RefreshTokenStore, s.app.KeyStore, s.app.Actives, s.app.Config, s.app.Reporter,
+		account.ID, &s.app.Config.ApplicationDomains[0], meta.GetRefreshToken(ctx),
+	)
+	if err != nil {
+		panic(err)
+	}
+
+	// Return the signed session in a metadata
+	meta.SetSession(ctx, s.app.Config.SessionCookieName, sessionToken)
+
+	// Return the signed identity token in the body
+	return &authnpb.LoginResponseEnvelope{
+		Result: &authnpb.LoginResponse{
+			IdToken: identityToken,
+		},
+	}, nil
+}
+
+func (s publicServer) RefreshSession(ctx context.Context, _ *authnpb.RefreshSessionRequest) (*authnpb.RefreshSessionResponseEnvelope, error) {
+
+	// check for valid session with live token
+	accountID := meta.GetSessionAccountID(ctx)
+	if accountID == 0 {
+		return nil, status.Error(codes.Unauthenticated, "account not found")
+	}
+
+	identityToken, err := services.SessionRefresher(
+		s.app.RefreshTokenStore, s.app.KeyStore, s.app.Actives, s.app.Config, s.app.Reporter,
+		meta.GetSession(ctx), accountID, &s.app.Config.ApplicationDomains[0],
```

Comment:
there are a few places where the REST server uses `route.MatchedDomain` to reflect the domain validated from an `Origin` or `Referer` header back as the `aud` claim. this is intended to support cross-domain logins, with an access token for each domain.

is it possible to support that with the gRPC server? it could matter to JavaScript clients that connect to gRPC through the REST interface.

Reply:
I probably can pull something together using the [`:authority`](https://github.com/grpc/grpc/blob/master/doc/PROTOCOL-HTTP2.md) header/metadata to emulate this. Let me look into it and experiment to figure something out.